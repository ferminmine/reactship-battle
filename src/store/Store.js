import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import { createBrowserHistory } from 'history';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import playerReducer from '../modules/player/PlayerReducer';
import shipReducer from '../modules/ship/ShipReducer';
import gameReducer from '../modules/game/GameReducer';
import shootsReducer from '../modules/shoots/ShootsReducer';

export const history = createBrowserHistory();

const allReducers = combineReducers({
  player: playerReducer,
  game: gameReducer,
  ships: shipReducer,
  shoots: shootsReducer,
  router: connectRouter(history)
});

export const ROOT_REDUCER_ACTION_TYPES = {
  RESET_APP: 'RESET_APP'
};
export const resetApp = () => ({ type: ROOT_REDUCER_ACTION_TYPES.RESET_APP });

const rootReducer = (state, action) => {
  if (action.type === 'RESET_APP') {
    state = undefined;
  }

  return allReducers(state, action);
};

const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default createStore(
  rootReducer,
  composeEnhancer(applyMiddleware(routerMiddleware(history)))
);
