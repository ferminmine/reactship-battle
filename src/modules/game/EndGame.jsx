import React from 'react';
import { compose, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { push } from 'connected-react-router';
import { resetApp } from '../../store/Store';
import GAME_STATES from '../game/GameStates';
import Swal from 'sweetalert2';

class EndGame extends React.Component {
  componentDidMount = () => {
    const { game } = this.props;
    let endGameMessage = {
      showCancelButton: false,
      confirmButtonText: 'I will start again!',
      allowOutsideClick: false
    };
    if (game.winner === 'player') {
      endGameMessage = {
        ...endGameMessage,
        title: 'Victory!',
        text: 'You won the game! Congratulations!',
        type: 'success'
      };
    }
    if (game.state === GAME_STATES.PLAYER_SURRENDERED) {
        endGameMessage = {
          ...endGameMessage,
          title: 'Surrender!',
          text: 'You chose to surrender! Are you up for a rematch?',
          type: 'error'
        };
      }
    if (game.winner === 'CPU') {
      endGameMessage = {
        ...endGameMessage,
        title: 'Defeat!',
        text: 'You lost the game! Are you up for a rematch?',
        type: 'error'
      };
    }
    this.throwEndGameMessage(endGameMessage);
  };

  throwEndGameMessage = async endGameMessage => {
    await Swal.fire(endGameMessage);
    this.resetGame();
  };

  resetGame = () => {
    this.props.resetApp();
    this.props.push('/');
  };

  render = () => {
    return <React.Fragment />;
  };
}

const mapStateToProps = state => ({
  game: state.game
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      push,
      resetApp
    },
    dispatch
  );

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(EndGame);
