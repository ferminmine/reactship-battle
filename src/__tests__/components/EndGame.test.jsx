import React from 'react';
import { mount } from 'enzyme';
import EndGame from '../../modules/game/EndGame';
import { Provider } from 'react-redux';
import store from '../../store/Store';
import renderer from 'react-test-renderer';

describe('EndGame Component Test', () => {
  it('renders without crashing', () => {
    mount(
      <Provider store={store}>
        {' '}
        <EndGame />{' '}
      </Provider>
    );
  });

  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <EndGame />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
