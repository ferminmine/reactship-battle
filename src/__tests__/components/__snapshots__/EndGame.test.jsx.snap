// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EndGame Component Test matches the snapshot 1`] = `null`;
